common/s3util.cpp

```diff
@@ -493,4 +507,124 @@ shared_ptr<S3Util> S3Util::BuildS3Util(
       new S3Util(bucket, aws_config, options, read_ratelimit_mb, write_ratelimit_mb));
 }
 
+
+S3Concurrent::S3Concurrent(const string &bucket, const int upload_MBps)
+  : bucket_(bucket),
+    upload_MBps_(upload_MBps),
+    executor_(2),  // n_threads
+    transfer_config_(&executor_) {
+
+  const uint64_t MiB = 1024*1024;
+  const uint64_t max_heap_size = 400 * MiB;
+  const uint64_t buffer_size = 25 * MiB;
+  int max_connections = 16; // upper limit, good for 10Gbps.
+
+  // throttle concurrency according to desired rate.
+  // 2 threads allows for some file interleaving.
+  // approx 70 MB/sec speed per connection.
+  if (upload_MBps_ != 0) {
+    max_connections = std::min(max_connections, std::max(1, (int)(upload_MBps_/70)));
+  }
+  LOG(INFO) << "S3Concurrent() upload_MBps: " << upload_MBps_
+            << " max_connections: " << max_connections
+            << " max_heap_size MiB: " << max_heap_size / MiB
+            << " buffer_size MiB: " << buffer_size / MiB;
+  int socket_timeout_ms = 3000;
+  client_config_.connectTimeoutMs = socket_timeout_ms;
+  client_config_.requestTimeoutMs = socket_timeout_ms;
+  client_config_.maxConnections = max_connections;
+  client_config_.region = "us-east-1";
+
+  if(s3_client_)
+    s3_client_.reset();
+  s3_client_ = std::make_shared<Aws::S3::S3Client>(client_config_);
+  transfer_config_.s3Client = s3_client_;
+  transfer_config_.transferBufferMaxHeapSize = max_heap_size;
+  transfer_config_.bufferSize = buffer_size;
+  transfer_config_.transferInitiatedCallback =
```

Comment:
will this be called for each task/file for the transfer manager?

Reply:
No, one S3Concurrent object (and therefore TransferManager) is created for the AdminHandler process, and is reused.